clash-prelude/src/Clash/Explicit/BlockRam.hs

```diff
@@ -1038,3 +1045,280 @@ readNew clk rst en ram rdAddr wrM = mux wasSame wasWritten $ ram rdAddr wrM
         (wasSame,wasWritten) =
           unbundle (register clk rst en (False, undefined)
                              (readNewT <$> rdAddr <*> wrM))
+
+-- | Produces vendor-agnostic HDL that will be inferred as a true, dual port
+-- block ram. It has write-first (or "new data") behavior, no clock enables,
+-- no byte enables, and is symmetric in its data ports.
+--
+-- * __NB__: Read value is delayed by 1 cycle
+-- * __NB__: Initial output value is /undefined/
+-- * __NB__: Port A and port B cannot make overlapping read/write requests on
+--           the same address.
+--
+trueDualPortBlockRam ::
+  forall nAddrs domA domB a .
+  ( HasCallStack
+  , KnownNat nAddrs
+  , KnownDomain domA
+  , KnownDomain domB
+  , NFDataX a
+  , BitPack a
+  ) =>
+
+  Clock domA ->
+  -- ^ Clock for port A
+  Signal domA (Maybe a) ->
+  -- ^ Write data port A
+  Signal domA (Index nAddrs) ->
+  -- ^ Address to read from or write to on port A
+
+  Clock domB ->
+  -- ^ Clock for port B
+  Signal domB (Maybe a) ->
+  -- ^ Write data B
+  Signal domB (Index nAddrs) ->
+  -- ^ Address to read from or write to on port B
+
+  (Signal domA a, Signal domB a)
+  -- ^ Outputs data on /next/ cycle. If write enable is @True@, the data written
+  -- will be echoed. If write enable is @False@, the read data is returned.
+
+trueDualPortBlockRam clkA writeA addrA clkB writeB addrB =
+  trueDualPortBlockRamWrapper
+    clkA (isJust <$> writeA) addrA (fromJustX <$> writeA)
+    clkB (isJust <$> writeB) addrB (fromJustX <$> writeB)
+
+toMaybeX :: a -> MaybeX a
+toMaybeX a =
+  case isX a of
+    Left _ -> IsX
+    Right _ -> IsDefined a
+
+data MaybeX a = IsX | IsDefined !a
+
+data Conflict = Conflict
+  { cfWrite :: !(MaybeX Bool)
+  , cfAddress :: !(MaybeX Int) }
+
+-- | "Stronger" conflict wins:
+--
+--   * Writes over read
+--   * Undefineds over anything
+--
+mergeConflicts :: Conflict -> Conflict -> Conflict
+mergeConflicts conflict1 conflict2 = Conflict
+  { cfWrite = mergeWrite (cfWrite conflict1) (cfWrite conflict2)
+  , cfAddress = mergeAddress (cfAddress conflict1) (cfAddress conflict2) }
+ where
+  mergeX _ IsX _ = IsX
+  mergeX _ _ IsX = IsX
+  mergeX f (IsDefined a) (IsDefined b) = IsDefined (f a b)
+
+  mergeWrite a b = mergeX (||) a b
+  mergeAddress a b = mergeX const a b
+
+-- | "Stronger" conflict wins:
+--
+--   * Writes over reads
+--   * Undefineds over reads/writes
+--   * @Just@s over @Nothing@s
+--
+mergeMaybeConflicts :: Maybe Conflict -> Maybe Conflict -> Maybe Conflict
+mergeMaybeConflicts Nothing Nothing = Nothing
+mergeMaybeConflicts (Just c) Nothing = Just c
+mergeMaybeConflicts Nothing (Just c) = Just c
+mergeMaybeConflicts (Just c1) (Just c2) = Just (mergeConflicts c1 c2)
```

Comment:
`mergeConflicts` looks to be a `Semigroup` instance for `Conflict`, and `mergeMaybeConflicts` is then just `<>` on `(Semigroup a) => Maybe a`